src/components/colorbar/connect.js

```diff
@@ -0,0 +1,85 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+
+var aggNums = require('../../lib').aggNums;
+var Colorscale = require('../colorscale');
+var drawColorbar = require('./draw');
+
+/**
+ * connectColorbar: create a colorbar from a trace, using its module to
+ *   describe the connection.
+ *
+ * @param {DOM element} gd
+ *
+ * @param {Array} cd
+ *   calcdata entry for this trace. cd[0].trace is the trace itself, and the
+ *   colorbar object will be stashed in cd[0].t.cb
+ *
+ * @param {object|function} moduleOpts
+ *   may be a function(gd, cd) to override the standard handling below. If
+ *   an object, should have these keys:
+ * @param {Optional(string)} moduleOpts.container
+ *   name of the container inside the trace where the colorbar and colorscale
+ *   attributes live (ie 'marker', 'line') - omit if they're at the trace root.
+ * @param {string} moduleOpts.min
+ *   name of the attribute holding the value of the minimum color
+ * @param {string} moduleOpts.max
+ *   name of the attribute holding the value of the maximum color
+ * @param {Optional(string)} moduleOpts.vals
+ *   name of the attribute holding the (numeric) color data
+ *   used only if min/max fail. May be omitted if these are always
+ *   pre-calculated.
+ */
+module.exports = function connectColorbar(gd, cd, moduleOpts) {
+    if(typeof moduleOpts === 'function') return moduleOpts(gd, cd);
```

Comment:
So, contour still gets its own version on `module._colorbar` in 

https://github.com/plotly/plotly.js/blob/fa8614750586a2901ccf65a520521bb570eb7026/src/traces/contour/colorbar.js

Reply:
... and that's the only exception?

Reply:
Yes, contour and those that inherit from it (`contourcarpet`, `contourgl`, `histogram2dcontour`).

I should note that there's one other place `module.colorbar` gets used, that's in [`colorbar.draw`](https://github.com/plotly/plotly.js/pull/2681/commits/fa8614750586a2901ccf65a520521bb570eb7026#diff-ad110e994c21bb045d288b9c96e4e6ccR452) for getting the `nestedProperty` string for title editing. If contour colorbars weren't at the module root we'd need to set a `container` property of the *function* which would be a bit hacky... alternatively we could either a) make this function be `module.colorbar.func`, or b) incorporate contour into this routine directly, triggered by some other property of `module.colorbar`.

Reply:
> alternatively we could either a) make this function be module.colorbar.func

This sounds cleaner to me. But no need to do this in this PR.